src/discrimination_tree.ml

```diff
@@ -0,0 +1,146 @@
+(* elpi: embedded lambda prolog interpreter                                  *)
+(* license: GNU Lesser General Public License Version 2.1 or later           *)
+(* ------------------------------------------------------------------------- *)
+
+module type IndexableTerm = sig
+  type cell
+  type path = cell list
+
+  val show_cell : cell -> string
+  val pp_cell : Format.formatter -> cell -> unit
+  val compare : cell -> cell -> int
+
+  (* You have h(f(x,g(y,z)),t) whose path_string_of_term_with_jl is
+     (h,2).(f,2).(x,0).(g,2).(y,0).(z,0).(t,0) and you are at f and want to
+     skip all its progeny, thus you want to reach t.
+
+     You need to skip as many elements as the sum of all arieties contained
+     in the progeny of f.
+
+     The input ariety is the one of f while the path is x.g....t
+     Should be the equivalent of after_t in the literature (handbook A.R.)
+  *)
+  (* MAYBE: a pointer to t from f should increase performances (i.e. jump list
+            from McCune 1990) *)
+  val skip : path -> path
+  val arity_of : cell -> int
+  val variable : cell
+  val pp : Format.formatter -> path -> unit
+  val show : path -> string
+end
+
+module type DiscriminationTree = sig
+  type data
+  type datalist = data list
+  type key
+  type keylist = key list
+  type t
+
+  include Elpi_util.Util.Show with type t := t
+
+  val iter : t -> (keylist -> datalist -> unit) -> unit
+  val fold : t -> (keylist -> datalist -> 'b -> 'b) -> 'b -> 'b
+  val empty : t
+  val index : t -> keylist -> data -> t
+  val remove_index : t -> keylist -> data -> t
+  val in_index : t -> keylist -> (data -> bool) -> bool
+  val retrieve_generalizations : t -> keylist -> datalist
+  val retrieve_unifiables : t -> keylist -> datalist
+end
+
+module type TimeStampList = sig
+  type elt
+  type t = elt list
+
+  include Elpi_util.Util.Show with type t := t
+
+  val get_time_stamp : elt -> int
+end
+
+module Make (K : IndexableTerm) (D : TimeStampList) :
+  DiscriminationTree
+    with type data = D.elt
+     and type datalist = D.t
+     and type key = K.cell
+     and type keylist = K.path = struct
+  module OrderedPathStringElement = struct
+    type t = K.cell
+
+    let show = K.show_cell
+    let pp = K.pp_cell
+    let compare = K.compare
+  end
+
+  module PSMap = Elpi_util.Util.Map.Make (OrderedPathStringElement)
+  module Trie = Trie.Make (PSMap)
+
+  type data = D.elt
+  type datalist = D.t
+  type key = K.cell
+  type keylist = K.path
+  type t = datalist Trie.t
+
+  let pp = Trie.pp D.pp
+  let show = Trie.show D.pp
+  let empty = Trie.empty
+  let iter dt f = Trie.iter (fun p x -> f p x) dt
+  let fold dt f = Trie.fold (fun p x -> f p x) dt
+
+  let index tree ps info =
+    let ps_set = try Trie.find ps tree with Not_found -> [] in
+    Trie.add ps (info :: ps_set) tree
```

Comment:
Maybe Trie.add could have this semantics  already (modulo having the trie be on 'a list all the times).